src/mbgl/renderer/render_layer.hpp

```diff
@@ -69,16 +79,23 @@ class RenderLayer {
 
     virtual std::unique_ptr<Bucket> createBucket(const BucketParameters&, const std::vector<const RenderLayer*>&) const = 0;
 
+    void setRenderTiles(std::vector<std::reference_wrapper<RenderTile>>);
     // Private implementation
     Immutable<style::Layer::Impl> baseImpl;
     void setImpl(Immutable<style::Layer::Impl>);
 
+    virtual const style::Layer::Impl& impl() const = 0;
```

Comment:
The virtual method's return type is overridden to the derived `Layer::Impl` type in each derived RenderLayer. That is needed so that `RenderLayer::uploadBuckets()` and `RenderLayer::render` can be called with the impl pointer referencing the derived class and not the base class.